mcp/src/main/java/io/modelcontextprotocol/server/McpAsyncServerExchange.java

```diff
@@ -101,4 +109,30 @@ public Mono<McpSchema.ListRootsResult> listRoots(String cursor) {
 				LIST_ROOTS_RESULT_TYPE_REF);
 	}
 
+	/**
+	 * Send a logging message notification to all connected clients. Messages below the
+	 * current minimum logging level will be filtered out.
+	 * @param loggingMessageNotification The logging message to send
+	 * @return A Mono that completes when the notification has been sent
+	 */
+	public Mono<Void> loggingNotification(LoggingMessageNotification loggingMessageNotification) {
+
+		if (loggingMessageNotification == null) {
+			return Mono.error(new McpError("Logging message must not be null"));
+		}
+
+		return Mono.defer(() -> {
+			if (loggingMessageNotification.level().level() < this.session.getMinLoggingLevel().level()) {
```

Comment:
Thie logic of checking should be internal to the entity that maintains the state and not leak into the caller (session should control and check it if the session has the set level, not the exchange which is just the user).

Reply:
it can